### PR TITLE
GameLogic

### PR DESCRIPTION
**Pull request by amiraelsayedbadwy**
check the selected difficulty according to  it set the level time
- Running a timer on background, stop counting when press Launch Parachute button
- animation on dots while page is counting if possible
- Calculate the difference between the timer and the LevelTime variable (changes depending on the level), when the difference is less than 1, show a blank popup
- If the difference is more than 1 second, then go to a blank LosePage
- Fix the pop ui
- Use Rg popup
- Show the animation
-When pressing next level button, dismisses the popup and increases the level number by 1
- Also store the time different in a list<int> so as to get the final average at the end of the game
- Use this animation instead of the black circle https://lottiefiles.com/3596-win-result-2

- Make sure that the next level increases the time by 1 second,
- The circles should also reflect the level number
- Make sure the timer is ready for this level

- If the user is on the level 10, then show Complete instead of Next Level for the popup and take them to the LandedPage
- Comes up after level 10!
- display this screen for 3 seconds and then hide it

- Finish the UI
- Hook up the buttons correctly
- Get the average from the List<int> that we had stored, and display it here
- Store the user’s score for the specific difficulty in a Preferences storage
### Testing
Android>home>start>choose difficulity >gameview>press launch>if greater than 1 >lose page
ios>home>start>choose difficulity >gameview>press launch>if greater than 1 >lose page

### CODE REVIEW

File: IslandLanding/IslandLanding/ViewModel/GameViewModel.cs [hunk 1]
```diff
@@ -1,30 +1,105 @@
...
+    public int LevelNumber { get; set; }
+    public DateTime StartedTime { get; set; }
     public ICommand ReadyCommand { get; set; }
     public ICommand BackCommand { get; set; }
     public ICommand RestartCommand { get; set; }
+    public ICommand LaunchCommand { get; set; }
     public ObservableCollection<LevelsModel> DotsList { get; set; }
-    public bool IsStarting{ get; set;}
+    public bool IsStarting { get; set; }
+    public GameModel gameModel { get; set; }
```
Comment (saamerm) [inline IslandLanding/IslandLanding/ViewModel/GameViewModel.cs:28]: Properties follow PascalCasing, please fix this @amiraelsayedbadwy

File: IslandLanding/IslandLanding/ViewModel/GameViewModel.cs [hunk 2]
```diff
@@ -1,30 +1,105 @@
...
     {
       ReadyCommand = new Command(ReadyCommandExcute);
       BackCommand = new Command(BackCommandExcute);
-      DotsList = new ObservableCollection<LevelsModel>();
+   
       RestartCommand = new Command(RestartCommandExcute);
+      LaunchCommand = new Command(LaunchCommandExcute);
+      TimeDifferencesList = new List<double>();
       DrawLevels();
+      MessagingCenter.Subscribe<NextPopupViewModel>(this,"nextlevel", (sender) =>
```
Comment (saamerm) [inline IslandLanding/IslandLanding/ViewModel/GameViewModel.cs:39]: Please use camelCasing for `nextlevel`

File: IslandLanding/IslandLanding/ViewModel/GameViewModel.cs [hunk 3]
```diff
@@ -1,30 +1,105 @@
...
+      {
+        IsStarting = false;
+        LevelNumber = Preferences.Get("levelnumber", 1) + 1;
```
Comment (saamerm) [inline IslandLanding/IslandLanding/ViewModel/GameViewModel.cs:42]: Please use camelCasing for `levelnumber`

File: IslandLanding/IslandLanding/ViewModel/GameViewModel.cs [hunk 4]
```diff
@@ -1,30 +1,105 @@
...
+          if(item.LevelNumber==LevelNumber.ToString())
+          {
+            item.IsCompleted = true;
+            item.BackgroundColor = Color.FromHex("#C5C5C5");
+          }
+          
+        }
+      });
+      //this part for check level and add level time according to choosen difficulity
+      if ((Preferences.Get("levelnumber", 1) == 1))
```
Comment (saamerm) [inline IslandLanding/IslandLanding/ViewModel/GameViewModel.cs:56]: Please use camelCasing for `levelnumber`

File: IslandLanding/IslandLanding/ViewModel/GameViewModel.cs [hunk 5]
```diff
@@ -41,29 +116,35 @@ private void BackCommandExcute(object obj)
     private void ReadyCommandExcute(object obj)
     {
       IsStarting = true;
-      CalculateTimeRemaining();
+      // store time .now in varaible 
+      // when press launch compare time.now with varaible 
+      // then sub the 2 times 
+      //CalculateTimeRemaining();
+      Preferences.Set("levelnumber", LevelNumber);
```
Comment (saamerm) [inline IslandLanding/IslandLanding/ViewModel/GameViewModel.cs:123]: Please use camelCasing for `levelnumber`

File: IslandLanding/IslandLanding/ViewModel/GameViewModel.cs [hunk 6]
```diff
@@ -41,29 +116,35 @@ private void BackCommandExcute(object obj)
     private void ReadyCommandExcute(object obj)
     {
       IsStarting = true;
-      CalculateTimeRemaining();
+      // store time .now in varaible 
+      // when press launch compare time.now with varaible 
+      // then sub the 2 times 
+      //CalculateTimeRemaining();
```
Comment (saamerm) [inline IslandLanding/IslandLanding/ViewModel/GameViewModel.cs:122]: Is CalculateTimeRemaining() still needed here?

File: IslandLanding/IslandLanding/ViewModel/GameViewModel.cs [hunk 7]
```diff
@@ -41,29 +116,35 @@ private void BackCommandExcute(object obj)
...
-          return true;
-        }
-        else
-        {
-          
-          Seconds = "0:0:0";
-          return false;
-        }
-      });
+      //DateTime timeExpired = DateTime.Now.AddSeconds(8);
```
Comment (saamerm) [inline IslandLanding/IslandLanding/ViewModel/GameViewModel.cs:131]: If you want to keep this, please add a Todo so if anyone has to pick this work, they know why it's commented out @amiraelsayedbadwy

File: IslandLanding/IslandLanding/ViewModel/NextPopupViewModel.cs [hunk 8]
```diff
@@ -0,0 +1,37 @@
...
+ public class NextPopupViewModel:BaseViewModel
+  {
+    public string Seconds { get; set; }
+    public string ButtonText { get; set; }
+    public ICommand NextCommand { get; set; }
+    public NextPopupViewModel( string difftime)
+    {
+      Seconds = difftime;
+      NextCommand = new Command(NextCommandExcute);
+      ButtonText = (Preferences.Get("levelnumber", 1) == 10) ? "Complet" : "Next";
```
Comment (saamerm) [inline IslandLanding/IslandLanding/ViewModel/NextPopupViewModel.cs:20]: Please use "Complete" instead of "Complet"

File: IslandLanding/IslandLanding/ViewModel/NextPopupViewModel.cs [hunk 9]
```diff
@@ -0,0 +1,37 @@
...
+using Xamarin.Forms;
+
+namespace IslandLanding.ViewModel
+{
+ public class NextPopupViewModel:BaseViewModel
+  {
+    public string Seconds { get; set; }
+    public string ButtonText { get; set; }
+    public ICommand NextCommand { get; set; }
+    public NextPopupViewModel( string difftime)
```
Comment (saamerm) [inline IslandLanding/IslandLanding/ViewModel/NextPopupViewModel.cs:16]: Please use camelCasing for `difftime`

File: IslandLanding/IslandLanding/ViewModel/NextPopupViewModel.cs [hunk 10]
```diff
@@ -0,0 +1,37 @@
...
+    {
+      Seconds = difftime;
+      NextCommand = new Command(NextCommandExcute);
+      ButtonText = (Preferences.Get("levelnumber", 1) == 10) ? "Complet" : "Next";
+
+    }
+
+    private void NextCommandExcute(object obj)
+    {
+      if (Preferences.Get("levelnumber", 1) > 10)
```
Comment (saamerm) [inline IslandLanding/IslandLanding/ViewModel/NextPopupViewModel.cs:26]: Please use camelCasing for levelnumber

File: IslandLanding/IslandLanding/ViewModel/NextPopupViewModel.cs [hunk 11]
```diff
@@ -0,0 +1,37 @@
...
+      {
+        // will navigate to win page 
+      }
+      else
+      {
+        MessagingCenter.Send<NextPopupViewModel>(this, "nextlevel");
```
Comment (saamerm) [inline IslandLanding/IslandLanding/ViewModel/NextPopupViewModel.cs:32]: Please use camelCasing for nextlevel

File: IslandLanding/IslandLanding/ViewModel/WinViewModel.cs [hunk 12]
```diff
@@ -0,0 +1,117 @@
...
+      {
+        IsWinning = false;
+
+        return false;
+      });
+      DrawLevels();
+    }
+    private void TryAginCommandExcute(object obj)
+    {
+      Preferences.Set("levelnumber", 1);
```
Comment (saamerm) [inline IslandLanding/IslandLanding/ViewModel/WinViewModel.cs:43]: Please use camelCasing for levelnumber

File: IslandLanding/IslandLanding/Views/LosePage.xaml [hunk 13]
```diff
@@ -0,0 +1,138 @@
...
+             VerticalTextAlignment="Center"
+             HorizontalTextAlignment="Center"
+             Margin="40,80,40,0" 
+             WidthRequest="313"
+             FontFamily="PatrickFont">
+        <Label.FormattedText>
+          <FormattedString>
+            <Span Text="Sorry , "/>
+            <Span Text="{Binding UserTag}"/>
+            <Span Text="You have landed on the lake."/>
```
Comment (saamerm) [inline IslandLanding/IslandLanding/Views/LosePage.xaml:35]: Update this to `<Span Text=". You have landed on the lake."/>`

File: IslandLanding/IslandLanding/Views/LosePage.xaml [hunk 14]
```diff
@@ -0,0 +1,138 @@
...
+             TextColor="White"
+             FontSize="30"
+             VerticalTextAlignment="Center"
+             HorizontalTextAlignment="Center"
+             Margin="40,80,40,0" 
+             WidthRequest="313"
+             FontFamily="PatrickFont">
+        <Label.FormattedText>
+          <FormattedString>
+            <Span Text="Sorry , "/>
```
Comment (saamerm) [inline IslandLanding/IslandLanding/Views/LosePage.xaml:33]: Update this to `<Span Text="Sorry, "/>`

File: IslandLanding/IslandLanding/Views/LosePage.xaml [hunk 15]
```diff
@@ -0,0 +1,138 @@
...
+             Margin="0,3,0,0"
+             VerticalTextAlignment="Center"
+             HorizontalTextAlignment="Center"
+             FontFamily="PatrickFont">
+          <Label.FormattedText>
+            <FormattedString>
+              <Span Text="It took you "/>
+              <Span  Text="{Binding gameModel.MainTime}"/>
+              <Span Text=" seconds, "/>
+              <Span Text="which is "/>
```
Comment (saamerm) [inline IslandLanding/IslandLanding/Views/LosePage.xaml:93]: Is there a reason we dont merge these two spans?
`<Span Text=" seconds, which is "/>`

File: IslandLanding/IslandLanding/Views/LosePage.xaml [hunk 16]
```diff
@@ -0,0 +1,138 @@
...
+
+        
+      
+
+      </StackLayout>
+      <StackLayout Grid.Row="3"  IsVisible="{Binding IsLosing,Converter={StaticResource InverseBool}}">
+        <Button 
+              WidthRequest="300"
+              Text="Main" Margin="{OnPlatform iOS='40,0,40,0', Android='40,-45,40,0'}"
+
```
Comment (saamerm) [inline IslandLanding/IslandLanding/Views/LosePage.xaml:108]: Unneeded line

File: IslandLanding/IslandLanding/Views/NextPopupPage.xaml [hunk 17]
```diff
@@ -0,0 +1,62 @@
...
+  xmlns:pancake="clr-namespace:Xamarin.Forms.PancakeView;assembly=Xamarin.Forms.PancakeView"
+  x:Class="IslandLanding.Views.NextPopupPage"
+  CloseWhenBackgroundIsClicked="True">
+  <AbsoluteLayout >
+    <pancake:PancakeView BackgroundColor="#898989"
+                                    AbsoluteLayout.LayoutFlags="PositionProportional,SizeProportional"
+                                     AbsoluteLayout.LayoutBounds="0,0,1,1"
+                                     VerticalOptions="FillAndExpand"
+                                     Opacity="0.3"/>
+
```
Comment (saamerm) [inline IslandLanding/IslandLanding/Views/NextPopupPage.xaml:16]: Why did you add 2 line breaks?

File: IslandLanding/IslandLanding/Views/NextPopupPage.xaml [hunk 18]
```diff
@@ -0,0 +1,62 @@
...
+            <Button Text="{Binding ButtonText}"
+                      WidthRequest="204"
+                      HeightRequest="40"
+                      CornerRadius="10"
+                      TextColor="White"
+                   HorizontalOptions="Center"
+                    Margin="0,10,0,0"
+                    Command="{Binding NextCommand}"
+                      BackgroundColor="{StaticResource orange}"/>
+
```
Comment (saamerm) [inline IslandLanding/IslandLanding/Views/NextPopupPage.xaml:43]: 2 line breaks again

File: IslandLanding/IslandLanding/Views/WinPage.xaml [hunk 19]
```diff
@@ -0,0 +1,137 @@
...
+             FontFamily="PatrickFont">
+        <Label.FormattedText>
+          <FormattedString>
+            <Span Text="{Binding UserTag}"/>
+            <Span Text=", You have successfully landed on the island!"/>
+          </FormattedString>
+        </Label.FormattedText>
+      </Label>
+      <pancake:PancakeView Grid.Row="1"
+
```
Comment (saamerm) [inline IslandLanding/IslandLanding/Views/WinPage.xaml:38]: What's this line break for?

File: IslandLanding/IslandLanding/Views/WinPage.xaml [hunk 20]
```diff
@@ -0,0 +1,137 @@
...
+        <Label Text="Your "
+             TextColor="White"
+             FontSize="40"
+             VerticalTextAlignment="Center"
+             HorizontalTextAlignment="Center"
+             FontFamily="PatrickFont"/>
+        <Label 
+             TextColor="White"
+             FontSize="30"
+
```
Comment (saamerm) [inline IslandLanding/IslandLanding/Views/WinPage.xaml:66]: What's this line break for?

File: IslandLanding/IslandLanding/Views/WinPage.xaml [hunk 21]
```diff
@@ -0,0 +1,137 @@
...
+              BorderColor="{StaticResource bordercolor}"
+              BorderWidth="4" 
+              HeightRequest="60"
+              CornerRadius="30"
+              FontSize="20"
+              FontFamily="PatrickFont"
+              BackgroundColor="{StaticResource orange}"
+              Command="{Binding TryAginCommand}"
+              />
+
```
Comment (saamerm) [inline IslandLanding/IslandLanding/Views/WinPage.xaml:106]: Why are there 3 line breaks here?

**Review by saamerm — CHANGES_REQUESTED** [review]
Please see the 21 comments, and respond to each one when you fix it

File: IslandLanding/IslandLanding/ViewModel/GameViewModel.cs (hunk 6 above)
Comment (amiraelsayedbadwy) [reply IslandLanding/IslandLanding/ViewModel/GameViewModel.cs:122]: no i just keep it

File: IslandLanding/IslandLanding/ViewModel/GameViewModel.cs (hunk 7 above)
Comment (amiraelsayedbadwy) [reply IslandLanding/IslandLanding/ViewModel/GameViewModel.cs:131]: ok

File: IslandLanding/IslandLanding/ViewModel/NextPopupViewModel.cs (hunk 8 above)
Comment (amiraelsayedbadwy) [reply IslandLanding/IslandLanding/ViewModel/NextPopupViewModel.cs:20]: ok

File: IslandLanding/IslandLanding/Views/LosePage.xaml (hunk 14 above)
Comment (amiraelsayedbadwy) [reply IslandLanding/IslandLanding/Views/LosePage.xaml:33]: ok

File: IslandLanding/IslandLanding/Views/WinPage.xaml (hunk 20 above)
Comment (amiraelsayedbadwy) [reply IslandLanding/IslandLanding/Views/WinPage.xaml:66]: done

File: IslandLanding/IslandLanding/ViewModel/GameViewModel.cs (hunk 1 above)
Comment (saamerm) [reply IslandLanding/IslandLanding/ViewModel/GameViewModel.cs:28]: gameModel should be GameModel

File: IslandLanding/IslandLanding/ViewModel/GameViewModel.cs (hunk 7 above)
Comment (saamerm) [reply IslandLanding/IslandLanding/ViewModel/GameViewModel.cs:118]: What's up with this code?

**Review by saamerm — CHANGES_REQUESTED** [review]
Why did you leave that code in there?

**Review by saamerm — APPROVED** [review]
